File: solutions/elixir/roman-numerals/1/lib/roman_numerals.ex
defmodule RomanNumerals do
  @numerals [
    [1000, "M"], [900, "CM"],
    [500, "D"], [400, "CD"],
    [100, "C"], [90, "XC"],
    [50, "L"], [40, "XL"],
    [10, "X"], [9, "IX"],
    [5, "V"], [4, "IV"],
    [1, "I"],
  ]

  def append_numeral([value, digit], [number, converted]) do
    [rem(number, value), converted <> String.duplicate(digit, div(number, value))]
  end
  
  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(number) do
    [_, converted] = List.foldl(@numerals, [number, ""], &append_numeral/2)
    converted
  end
end

File: solutions/julia/roman-numerals/1/roman-numerals.jl
const ROMAN_NUMERALS =
    [
        1000 => "M", 900 => "CM",
        500 => "D", 400 => "CD",
        100 => "C", 90 => "XC",
        50 => "L", 40 => "XL",
        10 => "X", 9 => "IX",
        5 => "V", 4 => "IV",
        1 => "I"
    ]

function replicateNumeral(number_converted, numeral)
    number = number_converted.first
    converted = number_converted.second

    to_add = numeral.second^div(number, numeral.first)

    return rem(number, numeral.first) => (converted * to_add)
end

function to_roman(number)
    if number == 0
        throw(ErrorException("Number can't be zero!"))
    elseif number < 0
        throw(ErrorException("Number can't be negative!"))
    end
    return foldl(replicateNumeral, ROMAN_NUMERALS; init=number => "").second
end
